### PR TITLE
Added tags and capabilities to model.

### PR DESCRIPTION
**Pull request by OSPFNeighbour**
Added tags and capabilities, expanded several other models to match up with Data Model POC

```json
{
  "data": {
    "user": {
      "organisation": {
        "name": "NSW SES"
      },
      "username": "chris",
      "email": "test@miceli.net.au",
      "devices": [
        {
          "uuid": "1234abc"
        }
      ],
      "tags": [
        {
          "name": "Geographical Area 1"
        }
      ],
      "capabilities": [
        {
          "name": "Capability 1"
        }
      ],
      "groups": [
        {
          "name": "Bankstown",
          "tags": [
            {
              "name": "Geographical Area 1"
            }
          ],
          "schedules": [
            {
              "name": "Bankstown Roster",
              "details": "Weekly roster",
              "deeplink": "https://foobar.com/"
            },
            {
              "name": "Bankstown Primary School Fete",
              "details": "5 people needed for school fete",
              "deeplink": "https://bankstownSES.com/forums/341231"
            }
          ],
          "events": [
            {
              "name": "Real Time Event #123",
              "details": "thing with the stuff",
              "deeplink": "https://foobar/event/123"
            }
          ]
        }
      ]
    }
  }
}
```

### CODE REVIEW

File: server/src/models.js [hunk 1]
```diff
@@ -64,6 +76,14 @@ UserModel.belongsTo(OrganisationModel);
 // devices belong to a single user
 DeviceModel.belongsTo(UserModel);
 
+//users -> tags
+UserModel.belongsToMany(TagsModel, { through: 'tag_user' });
```
Comment (sdunster) [inline server/src/models.js]: I think you want to swap these around:
```
TagsModel.belongsToMany(UserModel,{ through: 'user_tag' });
```
ultimately it might work out the same in the DB but I think it makes it easier to reason about. (the _user_ has the tag, not the _tag_ has the user)

File: server/src/models.js [hunk 2]
```diff
@@ -64,6 +76,14 @@ UserModel.belongsTo(OrganisationModel);
...
+
+//groups -> tags
+GroupModel.belongsToMany(TagsModel, { through: 'tag_group' });
```
Comment (sdunster) [inline server/src/models.js]: as above

File: server/src/models.js [hunk 3]
```diff
@@ -64,6 +76,14 @@ UserModel.belongsTo(OrganisationModel);
...
+
+//users belong to capabilities tags
+UserModel.belongsToMany(CapabilitiesModel, { through: 'capabilities_user' });
```
Comment (sdunster) [inline server/src/models.js]: as above

File: server/src/models.js [hunk 4]
```diff
@@ -103,25 +125,52 @@ db.sync({force: true}).then(() => {
...
           }),
-
           // create a device
           Device.create({
             uuid: '1234abc',
           }).then((device) => {
             device.setUser(user);
           }),
-
+          // create a capabilitie
```
Comment (sdunster) [inline server/src/models.js]: capability

File: server/src/models.js [hunk 5]
```diff
@@ -78,6 +98,8 @@ TimeSegmentModel.belongsTo(UserModel);
 const Organisation = db.models.organisation;
 const Group = db.models.group;
 const User = db.models.user;
+const Capabilities = db.models.capabilities;
```
Comment (sdunster) [inline server/src/models.js]: this should use the singular 'Capability'

File: server/src/models.js [hunk 6]
```diff
@@ -78,6 +98,8 @@ TimeSegmentModel.belongsTo(UserModel);
...
+const Tags = db.models.tags;
```
Comment (sdunster) [inline server/src/models.js]: as above 'Tag'

File: server/src/models.js [hunk 7]
```diff
@@ -27,6 +27,14 @@ const UserModel = db.define('user', {
   version: { type: Sequelize.INTEGER }, // version the password
 });
 
+const CapabilitiesModel = db.define('capabilities', {
```
Comment (sdunster) [inline server/src/models.js]: use singular `CapabilityModel` and `'capability'` for the table name

File: server/src/models.js [hunk 8]
```diff
@@ -27,6 +27,14 @@ const UserModel = db.define('user', {
...
+  name: { type: Sequelize.STRING },
+});
+
+const TagsModel = db.define('tags', {
```
Comment (sdunster) [inline server/src/models.js]: as above

**Comment by sdunster** [comment]
Oh one more thing... The tags are global to the system. Like everything else they really should be per-organisation so that organisation A would not see the tags that organisation B has defined.

**Comment by cmiceli** [comment]
So something which I think is worth considering from a code perspective, we might want to generalise the concept of a tag.

have a table of tags, and a table of associations (I'll call it TagAssociation) which have the concept of a scope (due to the non-unique id for tables). So the proposal is something like:

TagAssociation:
  - tag_id
  - scope (probably an enum which represents groups, users, etc)
  - object_id (id for the tagged object)

so with the scope and object_id we can now arbitrarily look everything up.

I think ultimately, a good solution would involve handing out unique id's across all tables but it might be a bit of over-engineering.

**Comment by OSPFNeighbour** [comment]
Ill address Sam's thing's first, for my own sake of self learning. Then delete everything and have a play on Chris' idea.

**Comment by sdunster** [comment]
I actually disagree with Chris's idea. I think it's much easier to just define seperate relationship tables to each of the things we want to attach tags to. One central tag table definitely makes sense though.
TL;DR I don't think you should change what you have now other than maybe unifying `Capability` with `Tag` although I don't think even that is worthwhile.

**Comment by OSPFNeighbour** [comment]
@sdunster regarding making them under the org, no idea how do that. flat DB i would store the id of the org against the tag... but dunno how do carry that concept over.

**Comment by OSPFNeighbour** [comment]
@sdunster deeplink is your term! you defined it on the data model.... a link back to the source of the request, for an event in the ses it would be a beacon link to the job.

**Comment by sdunster** [comment]
Re deeplink I assumed it was meant for locating the event or something. Lets just leave that stuff out and add it when we have a use for it.
Re making tags belong to an org - it's literally exactly that.
```
TagModel.belongsTo(OrganisationModel)
```

File: server/src/logic.js [hunk 9]
```diff
@@ -1,4 +1,4 @@
-import { Organisation, Group, User, Device, Event, Schedule, TimeSegment } from './models' ;
+import { Organisation, Group, User, Device, Event, Schedule, TimeSegment, Tags, Capabilities } from './models' ;
```
Comment (sdunster) [inline server/src/logic.js]: you didn't test this did you? `Capability` 👎

**Comment by OSPFNeighbour** [comment]
land this when ready if you are happy to expand to this.
